### DISCUSSION TITLE
운영체제- 스케줄링 알고리즘 - 서희 & 재찬

### DISCUSSION THREAD

**Discussion by LireEruel** [Shortudy-10th/tech-for-developer#128] [질의 응답]
- 스케쥴링이란 무엇이며, 스케쥴링이 필요한 이유는 무엇인가요?
- 선점 스케줄링과 비선점 스케줄링의 차이를 설명해주세요.
- 라운드 로빈은 어떤 원리로 동작하나요?
- 비선점형 스케줄링 알고리즘에 대해 설명해주세요.
- 선점형 스케줄링 알고리즘에 대해 설명해주세요.

**Comment by LireEruel** [comment]
### ❓️ 스케쥴링이란 무엇이며, 스케쥴링이 필요한 이유는 무엇인가요?

<br>

### 💬 답변
<blockquote>

**(운영체제) 스케줄링**
- 여러 프로세스나 스레드가 CPU의 시간을 공유하도록 관리하는 매커니즘
**스케줄링이 필요한 이유**
1. 효율성 증대
 - 여러 프로세스가 동시에 실행될 수 있으므로, 전체 시스템의 처리량과 응답시간을 개선
2. 공정성 보장
- 특정 프로세스가 다른 프로세스보다 더 많은 시간을 차지하지 않도록 관리
3. 우선순위 관리
- 중요한 작업이나 긴급한 작업에 더 많은 CPU 시간을 할당
4. 데드락 및 병목 현상 방지
- 스케줄링은 프로세스 간의 의존성 관리하여 데드락이나 병목 현상 퇴소화

</blockquote>

<br>

**Comment by LireEruel** [comment]
### ❓️ 선점 스케줄링과 비선점 스케줄링의 차이를 설명해주세요.

<br>

### 💬 답변
<blockquote>

**선점 스케줄링**
- 선점 스케줄링은 현재 실행 중인 프로세스가 완료되기 전에 다른 프로세스가 그 프로세스를 중단하고 CPU를 차지할 수 있는 스케줄링 방식
- 더 높은 우선 순위의 프로세스가 도착하면 현재 실행 중인 프로세스를 중단시키고 새로운 프로세스를 실행
- 시스템의 응답 시간을 개선하고, 더 공정한 자원 분배를 제공
- 실시간 운영체제, 대화형 시스템 등에서 주로 사용

**비선점 스케줄링 (Non-Preemptive Scheduling)**
- 비선점 스케줄링은 한 번 CPU를 할당받은 프로세스가 종료되기 전까지 다른 프로세스가 그 프로세스를 중단시킬 수 없는 스케줄링 방식
- 현재 실행 중인 프로세스가 완료될 때 까지 다른 프로세스가 대기 필요
- 구현이 상대적으로 쉬움
- 배치 처리 시스템, 간단한 인베디드 시스템에서 주로사용

</blockquote>

<br>

**Comment by kjc6735** [comment]
### ❓️ 라운드 로빈은 어떤 원리로 동작하나요?


<br>

### 💬 답변
<blockquote>

![image](https://github.com/Shortudy-10th/tech-for-developer/assets/56640518/d98b24b3-0d07-44a9-9f40-99782f726f87)


라운드 로빈 스케줄링은 대화형 시스템에서 사용되는 선점 형 스케줄링 방식이다. 

이 알고리즘 시간 할당량을 매 프로세스에게 주고 할당된 시간 안에 완료되지 못한 프로세스는 준비 큐의 맨 뒤에 배치시켜 cpu를 독점하지 않고 공평하게 이용할 수 있게 한다.


</blockquote>

<br>

**Comment by kjc6735** [comment]
### ❓️ 비선점형 스케줄링 알고리즘에 대해 설명해주세요.



<br>

### 💬 답변
<blockquote>
#### 비선점형 스케줄링
 - 한 프로세스가 CPU를 할당받으면 작업 종료 후 CPU 반환 시까지 다른 프로세스는 PCU 점유가 불가능한 스케줄링 방식
 - 모든 프로세스에 대한 요구를 공정하게 처리할 수 있지만, 짧은 작업을 수행하는 프로세스가 긴 작업 종료 시까지 대기해야할 수도 있다. (콘베이 현상)
 - 처리시간 편차가 적은 특정 프로세스 환경에 용이
</blockquote>

<br>

**Comment by kjc6735** [comment]
### ❓️ 선점형 스케줄링 알고리즘에 대해 설명해주세요.


<br>

### 💬 답변
<blockquote>

#### 선점형 스케줄링
 - 하나의 프로세스가 CPU를 차지하고 있을 때, 우선순위가 높은 다른 프로세스가 현재 프로세스를 중단시키고 CPU를 점유하는 스케줄링 방식
 - 비교적 응답이 빠르다는 장점이 있지만, 처리 시간을 예측하기 힘들고 높은 우선순위 프로세스들이 계속 들어오는 경우 오버헤드를 초래
 - 실시간 응답환경, Deadline 응답환경 등 우선순위가 높은 프로세스를 빠르게 처리해야 할 경우 등에 유용
</blockquote>

<br>
